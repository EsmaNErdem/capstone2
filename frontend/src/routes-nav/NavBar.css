/* NavBar.css */
.Navbar {
  background-color: #00000092;
  padding: 10px 20px;
  box-shadow: 2px 1px 3px yellowgreen;
  display: flex;
  justify-content: space-between;
  align-items: center;
  position: fixed; 
  top: 0; 
  width: 100%; 
  z-index: 1000;
}

.navbar-brand {
  text-decoration: none; 
  margin-left: 70px;
}

.navbar-brand .brand-text:hover {
  color: #ff6347; 
  transform: scale(1.1);
  transition: transform 0.3s ease;
}

.navbar-brand .brand-text {
  color: #a074f3; 
  font-size: 1.8rem;
  font-weight: bold;
  display: inline-block; 
}

.navbar-nav {
  display: flex;
  align-items: center;
  list-style: none;
  margin-right: 70px;
}

.nav-item {
  margin-right: 20px;
}

.nav-link {
  color: #36b9cc;
  font-size: 1.2rem;
  font-weight: bold;
  text-decoration: none;
  transition: color 0.3s;
}

.nav-link:hover {
  color: orangered; 
}

.avatar {
  cursor: pointer;
}

.dropdown-link {
  color: #36b9cc;
  text-decoration: none;
  font-size: 1.2rem;
  font-weight: bold;
}

.logout-link {
  color: #6d17b7;
  font-weight: bold;
}

/* Media Queries for Responsive Design */

@media (max-width: 768px) {
  .navbar-brand {
    margin-left: 0;
  }

  .Navbar {
    flex-direction: column;
  }

  .navbar-nav {
    margin-top: 10px;
    margin-right: 0;
    flex-direction: row;
  }

  .nav-item {
    margin-right: 10px;
    margin-bottom: 10px;
  }
}
