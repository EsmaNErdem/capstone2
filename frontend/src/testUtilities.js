import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  firstNsme: "testfirst",
  lastName: "testlast",
  email: "test@test.net",
  img: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => false, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const CurrUserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => true, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const NonUserProvider =
    ({ children, currentUser = null, hasLikedReview = () => false, hasLikedBook = () => false, hasFollowing = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing,}}>
      {children}
    </UserContext.Provider>
);

export { UserProvider, CurrUserProvider, NonUserProvider };
