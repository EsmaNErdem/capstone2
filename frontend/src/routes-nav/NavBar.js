import React, { useContext, useState } from "react";
import { Link, NavLink } from "react-router-dom";
import UserContext from "../auth/UserContext";
import { Menu, MenuItem, Avatar } from '@mui/material';
import "./NavBar.css";

/** Navbar bar for site. Shows up on every page.
 *
 * When the user is logged in, it shows links to the main areas of the site.
 * When not, it shows links to Login and Signup forms.
 *
 * Rendered by App.
 */
const NavBar = ({ logOut }) => {
  const { currentUser } = useContext(UserContext);
  const [anchorEl, setAnchorEl] = useState(null);

  const handleMenuOpen = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const handleMenuClose = () => {
    setAnchorEl(null);
  };

  function loggedInNav() {
    return (
      <ul className="navbar-nav ml-auto d-flex flex-row">
        <li className="nav-item mr-4">
          <NavLink className="nav-link" to="/reviews">
            Reviews
          </NavLink>
        </li>
        <li className="nav-item mr-4">
          <NavLink className="nav-link" to="/books">
            Books
          </NavLink>
        </li>
        <li className="nav-item mr-4">
          <NavLink className="nav-link" to="/chats">
            Chats
          </NavLink>
        </li>
        <li className="nav-item mr-4">
          <Avatar
            alt={currentUser.username}
            src={currentUser.img}
            className="avatar"
            onClick={handleMenuOpen}
          />
          <Menu
            anchorEl={anchorEl}
            open={Boolean(anchorEl)}
            onClose={handleMenuClose}
          >
            <MenuItem onClick={handleMenuClose}>
              <NavLink to={`/profile/${currentUser.username}`} className="dropdown-link">
                Profile
              </NavLink>
            </MenuItem>
            <MenuItem onClick={() => { logOut(); handleMenuClose(); }}>
              Log out {currentUser.first_name || currentUser.username}
            </MenuItem>
          </Menu>
        </li>
      </ul>
    );
  }

  function loggedOutNav() {
    return (
      <ul className="navbar-nav ml-auto d-flex flex-row">
        <li className="nav-item mr-4">
          <NavLink className="nav-link" to="/login">
            Login
          </NavLink>
        </li>
        <li className="nav-item mr-4">
          <NavLink className="nav-link" to="/signup">
            Sign Up
          </NavLink>
        </li>
      </ul>
    );
  }

  return (
    <nav className="Navbar navbar navbar-expand-md">
      <Link className="navbar-brand" to="/">
        <span className="brand-text">Book Chat</span>
      </Link>
      {currentUser ? loggedInNav() : loggedOutNav()}
    </nav>
  );
};

export default NavBar;
