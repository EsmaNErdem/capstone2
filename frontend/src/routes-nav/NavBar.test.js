import { MemoryRouter } from 'react-router-dom';
import { render, act, waitFor, fireEvent } from "@testing-library/react";
import NavBar from './NavBar';
import { UserProvider, CurrUserProvider, NonUserProvider } from "../testUtilities";

test("renders without crashing", () => {
    render(
    <MemoryRouter >
        <UserProvider>
          <NavBar />
        </UserProvider>
    </MemoryRouter>
    );
});

test("it renders and matches with snaphot while loggedin user", () => {
    const { asFragment } = render(
        <MemoryRouter >
            <UserProvider>
                <NavBar />
            </UserProvider>
        </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

test("it renders and matches with snaphot while no user loggedin", () => {
    const { asFragment } = render(
        <MemoryRouter >
            <NonUserProvider>
                <NavBar />
            </NonUserProvider>
        </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

test("Renders protected links while loggedin user", async () => {
    const { getByText, getByTestId } = render(
        <MemoryRouter >
            <CurrUserProvider>
                <NavBar />
            </CurrUserProvider>
        </MemoryRouter>
    );
    // "
    expect(getByText("Reviews")).toBeInTheDocument();
    expect(getByText("Books")).toBeInTheDocument();
    
    const profile = getByTestId('PersonIcon');
    await act(async () => {
        fireEvent.click(profile); 
    });
    await waitFor(() => {
        expect(getByText("Profile")).toBeInTheDocument();
        expect(getByText("Log out testuser")).toBeInTheDocument();
    }); 
});

test("Renders protected links while no loggedin user", () => {
    const { getByText } = render(
        <MemoryRouter >
            <NonUserProvider>
                <NavBar />
            </NonUserProvider>
        </MemoryRouter>
    );
    
    expect(getByText("Login")).toBeInTheDocument();
    expect(getByText("Sign Up")).toBeInTheDocument();
    expect(getByText("Book Chat")).toBeInTheDocument();
});