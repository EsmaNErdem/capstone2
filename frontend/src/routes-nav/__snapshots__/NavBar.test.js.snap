// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snaphot while loggedin user 1`] = `
<DocumentFragment>
  <nav
    class="Navbar navbar navbar-expand-md"
  >
    <a
      class="navbar-brand"
      href="/"
    >
      <span
        class="brand-text"
      >
        Book Chat
      </span>
    </a>
    <ul
      class="navbar-nav ml-auto d-flex flex-row"
    >
      <li
        class="nav-item mr-4"
      >
        <a
          class="nav-link"
          href="/reviews"
        >
          Reviews
        </a>
      </li>
      <li
        class="nav-item mr-4"
      >
        <a
          class="nav-link"
          href="/books"
        >
          Books
        </a>
      </li>
      <li
        class="nav-item mr-4"
      >
        <a
          class="nav-link"
          href="/chats"
        >
          Chats
        </a>
      </li>
      <li
        class="nav-item mr-4"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault avatar css-2s90m6-MuiAvatar-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
            data-testid="PersonIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </div>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;

exports[`it renders and matches with snaphot while no user loggedin 1`] = `
<DocumentFragment>
  <nav
    class="Navbar navbar navbar-expand-md"
  >
    <a
      class="navbar-brand"
      href="/"
    >
      <span
        class="brand-text"
      >
        Book Chat
      </span>
    </a>
    <ul
      class="navbar-nav ml-auto d-flex flex-row"
    >
      <li
        class="nav-item mr-4"
      >
        <a
          class="nav-link"
          href="/login"
        >
          Login
        </a>
      </li>
      <li
        class="nav-item mr-4"
      >
        <a
          class="nav-link"
          href="/signup"
        >
          Sign Up
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
